MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c

```diff
@@ -13,30 +13,119 @@ MU_CHANGE: new file
 #include <Library/DeviceStateLib.h>
 #include <Library/DebugLib.h>
 
+CHAR8  *DEVICE_STATE_STRINGS[6] = {
+  DEVICE_STATE_SECUREBOOT_OFF_STRING,
+  DEVICE_STATE_MANUFACTURING_MODE_STRING,
+  DEVICE_STATE_DEVELOPMENT_BUILD_ENABLED_STRING,
+  DEVICE_STATE_SOURCE_DEBUG_ENABLED_STRING,
+  DEVICE_STATE_UNDEFINED_STRING,
+  DEVICE_STATE_UNIT_TEST_MODE_STRING,
+};
+
 /**
-Function to Get current device state
-@retval the current DEVICE_STATE
-**/
-DEVICE_STATE
+ * Get String that represents the device's current insecure states, using the platform defined set of
+ * insecure device states. This can be used to measure insecure device states into the TPM or perform
+ * other required platform actions when the device enters an insecure state.
+ *
+ * @param[out] Buffer - Buffer to store the string
+ * @param[in] MaxSize - Maximum size of the buffer
+ * @retval The The number of characters in the string, including the null character.
+ */
+UINTN
 EFIAPI
-GetDeviceState (
+GetInsecureDeviceStateString (
+  OUT CHAR8  *Buffer,
+  IN UINTN   MaxSize
   )
 {
-  DEVICE_STATE  DevState = PcdGet32 (PcdDeviceStateBitmask);
+  CHAR8         *String;
+  DEVICE_STATE  BitmaskFromIndex;
+  DEVICE_STATE  CurrentDeviceState         = GetDeviceState ();
+  DEVICE_STATE  InsecureDeviceStateSetting = GetInsecureDeviceStateSetting ();
+
+  UINT32  NumberOfDeviceStates = sizeof (DEVICE_STATE_STRINGS) / sizeof (DEVICE_STATE_STRINGS[0]);

```

Comment:
If there is not a compelling reason for `NumberOfDeviceStates` to be a `UINT32`, it should be a `UINTN`.

Reply:
Ha, for this and the loop index below Vivian had them as UINTN and I said if there isn’t a compelling reason to have them UINTN, they should be a fixed type. I think we see a lot of weirdness around UINTN, generally not for a good reason. But, I defer to your judgement

Reply:
Yeah, `UINTN` is generally best for counts and indices because those tend to get used in pointer arithmetic and represent values that scale with pointer width. For example, `UINTN` on a 64-bit machine can accurately represent the number of bytes for a valid offset where that offset can be larger than `MAX_UINT32`.

There are of course cases where smaller integers make sense. Especially in firmware, we have fixed size fields in data structures irrespective of the current architecture. In that case, the maximum allowable value is constrained by the space allocated for the data rather than the architecture.

In both cases, it can simplify casting and result in overall better integer handling to apply the appropriate type for the use case. In this case, I see count and indexing into an array being used which is classically better handled with `UINTN`. And you usually need the count and index type to match for their comparison in the loop to be safe.